File: src/kitty.rs
use crate::encode::EncodedImage;
use std::fs::OpenOptions;
use std::io::{Write, BufWriter};

pub fn transmit_and_place(img: &EncodedImage, _image_id: u32, col: u16, row: u16) {
    let tty = OpenOptions::new()
        .write(true)
        .open("/dev/tty")
        .expect("cannot open /dev/tty");
    let mut out = BufWriter::new(tty);

    // position cursor
    write!(out, "\x1b[{};{}H", row, col).unwrap();

    let total = img.chunks.len();
    for (i, chunk) in img.chunks.iter().enumerate() {
        let is_last = i == total - 1;
        let m = if is_last { 0 } else { 1 };

        if i == 0 {
            write!(
                out,
                "\x1b_Ga=T,f=100,C=1,q=2,m={};{}\x1b\\",
                m, chunk
            ).unwrap();
        } else {
            write!(out, "\x1b_Gm={};{}\x1b\\", m, chunk).unwrap();
        }
    }

    out.flush().unwrap();
}

File: lua/mobius/kitty.lua
local M = {}

local function get_outer_tty()
    local pid = vim.fn.getpid()
    local path = vim.fn.system("readlink /proc/" .. pid .. "/fd/0"):gsub("%s+", "")
    return path
end

local function send_to_kitty(sequence)
    local tty_path = get_outer_tty()
    local f = io.open(tty_path, "wb")
    if not f then
        vim.notify("[mobius] cannot open tty: " .. tty_path, vim.log.levels.WARN)
        return
    end
    f:write(sequence)
    f:flush()
    f:close()
end

function M.get_cell_size()
    return 14, 28
end

function M.send(png_path, col, row)
    local f = io.open(png_path, "rb")
    if not f then return end
    local bytes = f:read("*a")
    f:close()

    local b64 = vim.base64.encode(bytes)
    local seq = string.format("\x1b[%d;%dH", row, col)

    local chunk_size = 4096
    local total = #b64
    local pos = 1
    local first = true

    while pos <= total do
        local chunk = b64:sub(pos, pos + chunk_size - 1)
        pos = pos + chunk_size
        local is_last = pos > total
        local m = is_last and 0 or 1

        if first then
            seq = seq .. string.format(
                "\x1b_Ga=T,f=100,C=1,q=2,m=%d;%s\x1b\\", m, chunk
            )
            first = false
        else
            seq = seq .. string.format("\x1b_Gm=%d;%s\x1b\\", m, chunk)
        end
    end

    send_to_kitty(seq)
end

function M.delete(image_id)
    send_to_kitty(string.format("\x1b_Ga=d,d=i,i=%d\x1b\\", image_id))
end

return M
